Bom weather platform

Zac-HD opened the pull request:
**Description:**

I've taken advantage of the lovely new `weather` component to simplify (accurate) weather observations for Aussies.  This is basically a thin wrapper around the `sensor.bom` component, so it doesn't have much in the way of new documentation.  Tests are identical to `weather.openweathermap`, which is to say absent - but all the functional parts are tested as part of the sensor.

Corresponds with home-assistant/home-assistant.github.io#1703, which includes example YAML configuration.

**Checklist:**

If user exposed functionality or configuration variables are added/changed:
  - [x] Documentation added/updated in [home-assistant.github.io](https://github.com/home-assistant/home-assistant.github.io)

If the code communicates with devices, web services, or third-party tools:
  - [x] Local tests with `tox` run successfully. **Your PR cannot be merged unless tests pass**
  - [x] New files were added to `.coveragerc`.

Zac-HD commented on homeassistant/components/sensor/bom.py:
```diff
@@ -164,19 +163,19 @@ def _build_url(self):
-             datetime.timedelta(minutes=35)):
+        curr_time = datetime.datetime.now(datetime.timezone.utc)
+        if self._lastupdate is not None and \
+            ((curr_time - self._lastupdate) < datetime.timedelta(minutes=30)):
```
Less inelegant than breaking the condition or terse names.  Happy to change if a human disagrees with me :smile:

Danielhiversen replied:
The tests have to pass before we will merge: https://travis-ci.org/home-assistant/home-assistant/jobs/188492332
So you have fix the indention.

Zac-HD replied:
No problem.

Zac-HD commented on homeassistant/components/weather/bom.py:
```diff
@@ -0,0 +1,99 @@
+                      'S', 'SSW', 'SW', 'WSW',
+                      'W', 'WNW', 'NW', 'NNW']
+        return {name: idx * 360 / 16 for idx, name in enumerate(directions)
+               }.get(self.bom_data.data.get('wind_dir'))
```
Oh come on.

Zac-HD commented:
[This](https://travis-ci.org/home-assistant/home-assistant/jobs/188492332#L231) is [simply wrong](https://github.com/home-assistant/home-assistant/pull/5153/files#diff-R168)

Zac-HD commented:
OK!  With a combination of minor style changes and just repeating the lint jobs, everything is passing now. :smile:

fabaff commented on homeassistant/components/weather/bom.py:
```diff
@@ -0,0 +1,99 @@
+        stationname = self.bom_data.data.get('name')
+        if stationname:
+            return 'BOM station {}'.format(stationname)
+        return 'Unknown BOM station'
```
I think that we should allow the user to set a custom name.

fabaff commented on homeassistant/components/weather/bom.py:
```diff
@@ -0,0 +1,99 @@
+    @property
+    def temperature_unit(self):
+        """Return the unit of measurement."""
+        raise TEMP_CELSIUS
```
Shouldn't this be a return?

Zac-HD replied:
Whoops, it should indeed.

Zac-HD commented:
This is probably a good time to mention - last night I worked out how to find the closest station with a zone_id and wmo_id, which would make all the config beyond `weather: {platform: bom}` optional :smile:

However that's likely to take a few days to hammer into shape, and change the docs etc in a different way, so I think it's better off as a separate pull request.

Zac-HD committed:
Fix typo

Zac-HD committed:
Auto-config for `sensor.bom`

Deprecate (but still support) the old two-part station ID, and move to a
single `station` identifier.  Any combination of these, including none,
is valid; most results in downloading and caching the station map to
work out any missing info.

Zac-HD committed:
Add `weather.bom` platform

Very similar to `sensor.bom`, but supporting the lovely new `weather`
component interface.  Easier to configure, and does not support the
deprecated config options.

Zac-HD commented:
OK, I've rebased and squashed to a pleasing history.  I also added a function to find the closest station automatically, making configuration entirely optional 😄

`sensor.bom` will continue to support the old two-part config (`zone_id` and `wmo_id`), on a best-effort basis - both is equivalent to providing `station`, `wmo_id` is unique but requires fetching the station map as for autoconfiguration, and `zone_id` alone is useless so we log and fall back to autodetection.  `weather.bom` only supports `station` and `name` config options.

Zac-HD commented:
@fabaff - I'm finished here.  Are there any changes you want before merging?

fabaff commented on homeassistant/components/weather/bom.py:
```diff
@@ -0,0 +1,102 @@
+    def update(self):
+        """Update current conditions."""
+        self.bom_data.update()
+
```
`def condition(self):` will raise an `NotImplementedError` execption.

Zac-HD replied:
What is this meant to return?  It looks like a short textual description, which unfortunately is not supplied by the BOM (well, actually it returns `"weather": "-"`...).  As a fallback, is it more appropriate to return an empty string or None?

Zac-HD replied:
Fixed.  From a few manual checks this will often return `'-'`.  Would it be better to check and return `None` (or `''`?) in such cases?

fabaff commented on homeassistant/components/sensor/bom.py:
```diff
@@ -66,25 +74,53 @@
+    vol.Optional(CONF_ZONE_ID): cv.string,
+    vol.Optional(CONF_WMO_ID): cv.string,
     vol.Optional(CONF_NAME, default=None): cv.string,
+    vol.Optional(CONF_STATION): validate_station,
```
`vol.Inclusive` and `vol.Exclusive` can help to distinguish between optional configuration values.

Zac-HD replied:
Hmm.  I've put `CONF_ZONE_ID` and `CONF_WMO_ID` in an inclusive group, so it's valid to provide both or neither but not one - this does simplify things and is more correct than optional for both.

However I can't work out how to express 'A xor (B and C)' - mixing inclusive and exclusive seems doomed to false-positives or false-negatives in every combination I've tried.  So I'm just logging a warning in `platform_setup`, and removing the documentation of the old style should avoid new users of it.

fabaff commented on homeassistant/components/sensor/bom.py:
```diff
@@ -66,25 +74,53 @@
+    if station is not None:
+        pass
+    elif zone_id and wmo_id:
+        station = zone_id + '.' + wmo_id
```
It would be preferred it string formatting is used. Also, in the rest of the file.

Zac-HD replied:
Done.

fabaff commented on homeassistant/components/sensor/bom.py:
```diff
@@ -182,3 +217,69 @@ def update(self):
+                    if wmo != '..':
+                        latlon[wmo] = (float(lat), float(lon))
+    zones = {}
+    pattern = (r'<a href="/products/(?P<zone>ID[DNQTVW]60801)/'
```
How about different area? 60901?

Zac-HD replied:
Unfortunately not listed on the given pages when I checked, but I'll generalise this and the other check to accept any valid product ID instead of just those I found.

fabaff commented on homeassistant/components/sensor/bom.py:
```diff
@@ -66,25 +74,53 @@
+    rest = BOMCurrentData(hass, station)
+    sensors = [BOMCurrentSensor(rest, variable, config.get(CONF_NAME))
+               for variable in config[CONF_MONITORED_CONDITIONS]]
     try:
```
I'm guessing...wouldn't it be more efficient to check the data object before we are creating the sensors?

Zac-HD replied:
Done.

Zac-HD commented on homeassistant/components/sensor/bom.py:
```diff
@@ -66,25 +74,53 @@
+    if station is None:
+        return
+    station = station.replace('.shtml', '')
+    if not re.fullmatch(r'ID[DNQTVW]60801\.\d\d\d\d\d', station):
```
We won't autodetect such stations I think, so it's particularly important to allow manual config!  Fixed.

Zac-HD committed:
Review improvements to BOM weather

Largely around better input validation.

Zac-HD commented:
@fabaff - anything else?

Zac-HD commented:
@fabaff - anything else?

Zac-HD commented:
Ping @fabaff - I've made all the changes you requested.  Is there anything else I need to do, or can you merge this so it's in before 0.36?
